### ISSUE TITLE
Test host process crashes on unhandled exceptions.

### ISSUE THREAD

**Issue by cpopescu89** [xunit/xunit#2992]
Xunit version - 2.8.0 / 2.9.0
dotnet 8

Recently, after updating to latest Xunit version (starting 2.8.0), the test host process started crashing.
Apparently the problem appears when there are multiple tests being run in parallel and unhandled exceptions are being thrown.
In this case (see the project that I attached), the problem occurs because the HTTPClient MyClient seems to not be initialized correctly because it's declared again in Class 6.
By using the Xunit 2.4.2, the crash does not occur, MyClient being initialized correctly even though it's declared again in class 6. 

It's also crashing on other instances as well in my main project, sometimes when the API returns 404/503 or unexpected errors, but this is at the moment the only way that I found to reproduce the issue that I'm seeing.

[XunitTestReport.zip](https://github.com/user-attachments/files/16470335/XunitTestReport.zip)

**Comment by bradwilson** [comment]
The fact that this *isn't* crashing in 2.4.2 seems more like a bug than the fact that it *is* crashing in 2.9.0.

The call stack for the failure is quite clear:

```
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at XunitTestReport.ChildClass.Dispose() in C:\Dev\repro\xunit2992\XunitTestReport\ChildClass.cs:line 11
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
```

There is no xUnit.net code involved here at all. The error is being thrown by `ChildClass`, because it's incorrectly trying to call Dispose() on a a null `Client` object.

In fact, the compiler quite clearly tells that there are three unresolved issues here:

```
C:\Dev\repro\xunit2992\XunitTestReport\Class6.cs(8,24): warning CS0108: 'Class6.Client' hides inherited member 'ChildClass.Client'. Use the new keyword if hiding was intended. [C:\Dev\repro\xunit2992\Xuni
tTestReport\XunitTestReport.csproj]
C:\Dev\repro\xunit2992\XunitTestReport\ChildClass.cs(7,28): warning CS8618: Non-nullable property 'Client' must contain a non-null value when exiting constructor. Consider declaring the property as nullab
le. [C:\Dev\repro\xunit2992\XunitTestReport\XunitTestReport.csproj]
C:\Dev\repro\xunit2992\XunitTestReport\Class6.cs(8,24): warning CS8618: Non-nullable property 'Client' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
[C:\Dev\repro\xunit2992\XunitTestReport\XunitTestReport.csproj]
```

First, `Class6` redeclares `Client`, and setting that `Client` does not set the `Client` property in `ChildClass`.

Second, there is no guarantee that `ChildClass.Client` is not null (and in fact you end up with a null reference exception precisely because it *is* null).

Third, there is no guarantee that `Class6.Client` is not null, either (though that does not appear to be a problem here).

The null reference exception is a combination of the first two compiler warnings being correct, plus `ParentClass.InitializeProperties` calling `GetType().GetProperties()` to initialize the properties on the test class. In the case of `Class6`, it sets the `Class6` version of the `Client` property but the `ChildClass` property remains unset. The exception is thrown when `Class6` is being cleaned up, because the `Dispose` method it inherits from `ChildClass` is being called, and the `ChildClass.Client` property was never set: a null reference exception ensues.

The reason this is getting surfaced as a host failure instead of a failed test is because the implementation of `ChildClass.Dispose` is marked as `async void`. This is where your bug got thrown into "test host crashing" instead of "test failing", because `async void` for `Dispose` is not supported. In practical terms it means _nobody is waiting for the results of the `Dispose` method_, because it's being run on a background thread. This is a really good way for errors to occur that go unnoticed, which is what's happening in 2.4.2. _Your bug is still there in 2.4.2, but it's being ignored._

If you had instead overrode `DisposeAsync` from `ParentClass` like this:

```csharp
public override async Task DisposeAsync()
{
    await Client.GetRandomCatFacts();
    await base.DisposeAsync();
}
```

Now when you run the tests, since the framework is waiting for the result of the `DisposeAsync` call (because it's `async Task`, not `async void`), you'll see that instead all three of your `Class6` tests are now are failing:

```
[xUnit.net 00:00:00.61]     XunitTestReport.Class6.TestMethod2(test: "test3") [FAIL]
  Failed XunitTestReport.Class6.TestMethod2(test: "test3") [457 ms]
  Error Message:
   System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at XunitTestReport.ChildClass.DisposeAsync() in C:\Dev\repro\xunit2992\XunitTestReport\ChildClass.cs:line 9
[xUnit.net 00:00:02.94]     XunitTestReport.Class6.TestMethod2(test: "test") [FAIL]
  Failed XunitTestReport.Class6.TestMethod2(test: "test") [2 s]
  Error Message:
   System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at XunitTestReport.ChildClass.DisposeAsync() in C:\Dev\repro\xunit2992\XunitTestReport\ChildClass.cs:line 9
[xUnit.net 00:00:03.25]     XunitTestReport.Class6.TestMethod2(test: "test2") [FAIL]
  Failed XunitTestReport.Class6.TestMethod2(test: "test2") [315 ms]
  Error Message:
   System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at XunitTestReport.ChildClass.DisposeAsync() in C:\Dev\repro\xunit2992\XunitTestReport\ChildClass.cs:line 9

Failed!  - Failed:     3, Passed:    18, Skipped:     0, Total:    21, Duration: 2 s - XunitTestReport.dll (net8.0)
```

-----

To summarize:

1. Don't use `async void` with `Dispose`;
2. You probably shouldn't ignore compiler warnings;
3. The test host crashing is in my mind perfect reasonable behavior.

**Comment by bradwilson** [comment]
Closing as "by design".
